File: src/main/scala/com/vc/hard/L1163.scala
package com.vc.hard

object L1163 {
  def lastSubstring(s: String): String = {
    import scala.collection.mutable
    var maxChar = '@'
    var maxCharIndexes = new mutable.HashSet[Int]()
    var i = 0

    while(i < s.length) {
      val ch = s(i)
      if(maxChar < ch) {
        maxChar = ch
        maxCharIndexes.clear()
      }
      if(maxChar == ch) maxCharIndexes += i
      while(i < s.length - 1 && s(i) == s(i + 1)) i += 1
      i += 1
    }

    var shift = 1
    while(maxCharIndexes.size != 1) {
      val nextLevel = new mutable.HashSet[Int]()
      val toDelete = new mutable.HashSet[Int]()
      var maxChar = '@'
      maxCharIndexes.foreach(index => {
        val newIndex = index + shift
        if(newIndex < s.length && !toDelete.contains(newIndex)) {
          if(maxChar < s(newIndex)) {
            maxChar = s(newIndex)
            nextLevel.clear
          }
          if(maxChar == s(newIndex)) {
            nextLevel += index
          }
          if(maxCharIndexes.contains(newIndex)) {
            toDelete += newIndex
          }
        }
      })
      toDelete.foreach(delIndex => {
        nextLevel.remove(delIndex)
      })
      maxCharIndexes = nextLevel
      shift += 1
    }
    s.substring(maxCharIndexes.head)
  }
}

File: src/main/java/com/vc/hard/L1163.java
package com.vc.hard;

class L1163 {
    public String lastSubstring(String s) {
        int i = 0, j = 1, offset = 0;
        int len = s.length();
        while(i + offset < len && j + offset < len) {
            char c = s.charAt(i + offset);
            char d = s.charAt(j + offset);
            if(c == d) offset++;
            else {
                if(c < d) i += offset + 1;
                else j += offset + 1;
                if(i == j) i++;
                offset = 0;
            }
        }
        return s.substring(Math.min(i, j));
    }
}